cmd/builder/internal/builder/main.go

```diff
@@ -139,21 +179,37 @@ func GetModules(cfg Config) error {
 		return fmt.Errorf("failed to update go.mod: %w", err)
 	}
 
-	cfg.Logger.Info("Getting go modules")
-	// basic retry if error from go mod command (in case of transient network error). This could be improved
-	// retry 3 times with 5 second spacing interval
-	retries := 3
-	failReason := "unknown"
-	for i := 1; i <= retries; i++ {
-		if err := runGoCommand(cfg, "mod", "download"); err != nil {
-			failReason = err.Error()
-			cfg.Logger.Info("Failed modules download", zap.String("retry", fmt.Sprintf("%d/%d", i, retries)))
-			time.Sleep(5 * time.Second)
+	if cfg.SkipStrictVersioning {
+		return downloadModules(cfg)
+	}
+
+	// Perform strict version checking.  For each component listed and the
+	// otelcol core dependency, check that the enclosing go module matches.
+	mf, mvm, err := cfg.readGoModFile()
+	if err != nil {
+		return err
+	}
+
+	coremod, corever := cfg.coreModuleAndVersion()
+	if mvm[coremod] != corever {
+		return fmt.Errorf("core collector version calculated by component dependencies %q does not match configured version %q: %w", mvm[coremod], corever, ErrStrictMode)
+	}
+
+	for _, mod := range cfg.allComponents() {
+		module, version, _ := strings.Cut(mod.GoMod, " ")
+		if module == mf.Module.Path {
+			// Main module is not checked, by definition.  This happens
+			// with --skip-new-go-module where the enclosing module
+			// contains the components used in the build.
 			continue
 		}
```

Comment:
Is this only relevant with #9631?

Reply:
I updated the comment here (and variable names).  It shouldn't be possible to hit this if statement without `--skip-new-go-module`, but I would like to keep this if statement here if that's okay. 🙂 If not, I can remove it.